CONTRIBUTING.md

```diff
@@ -34,9 +34,116 @@ The Python code in garage conforms to the [PEP8](https://www.python.org/dev/peps
+    
+* Prefer single-quoted strings (`'foo'`) over double-quoted strings (`"foo"`).
+
+    Double-quoted strings can be used if there is a compelling escape or formatting reason for using single quotes (e.g. a single quote appears inside the string).
+    
+* Add convenience imports in `__init__.py` of a package for shallow first-level repetitive imports, but not for subpackages, even if that subpackage is defined in a single `.py` file.
+
+    For instance, if an import line reads `from garage.foo.bar import Bar` then you should add `from garage.foo.bar import Bar` to `garage/foo/__init__.py` so that users may instead write `from garage.foo import Bar`. However, if an import line reads `from garage.foo.bar.stuff import Baz`, *do not* add `from garage.foo.bar.stuff import Baz` to `garage/foo/__init__.py`, because that obscures the `stuff` subpackage.
+
+    *Do*
+    
+    `garage/foo/__init__.py`:
+    ```python
+    """Foo package."""
+    from garage.foo.bar import Bar
+    ```
+    `garage/barp/bux.py`:
+    ```python
+    """Bux tools for barps."""
+    from garage.foo import Bar
+    from garage.foo.stuff import Baz
+    ```
+    
+    *Don't*
+    
+    `garage/foo/__init__.py`:
+    ```python
+    """Foo package."""
+    from garage.foo.bar import Bar
+    from garage.foo.bar.stuff import Baz
+    ```
+    `garage/barp/bux.py`:
+    ```python
+    """Bux tools for barps."""
+    from garage.foo import Bar
+    from garage.foo import Baz
+    ```
+    
+* Base and interface classes (i.e. classes which are not intended to ever be instantiated) should use the `abc` package to declare themselves as abstract.
+
+   i.e. your class should inherit from `abc.ABC` or use the metaclass `abc.ABCMeta`, it should declare its methods abstract (e.g. using `@abc.abstractmethod`) as-appropriate. Abstract methods should all use `pass` as their implementation, not `raise NotImplementedError`
+   
+   *Do*
+   ```python
+   import abc
+   
+   class Robot(abc.ABC):
+       """Interface for robots."""
+       
+       @abc.abstractmethod
+       def beep(self):
+           pass
+    ```
+    
+    *Don't*
+    ```python
+    
+    class Robot(object):
+        "Base class for robots."""
+        
+        def beep(self):
+            raise NotImplementedError
+    ```
+
+* When using external dependencies, use the `import` statement only to import whole modules, not individual classes or functions.
+
+    This applies to both packages from the standard library and 3rd-party dependencies. If a package has a long or cumbersome full path, or is used very frequently (e.g. `numpy`, `tensorflow`), you may use the keyword `as` to create a file-specific name which makes sense.
+    
+    *Do*
+    ```python
+    import collections
+    
+    import gym.spaces
+    import tensorflow as tf
+    
+    from garage.tf.models import MLPModel
```

Comment:
Why shouldn't we do `import garage.tf.models` here and do `garage.tf.models.MLPModel` in code? Is it because it's too long?

Reply:
The purpose of this rule is to make it easy to see anywhere in the code where we're using external vs internal dependencies.

Forcing people to write out long names for external deps makes it obvious that something is from an external dependency, even if the import is 1000 lines up, e.g. `q = collections.deque(10)` vs `q = deque(10)`.

```python
import collections

from garage.tf.model import MLPModel

q = collections.deque(10)
m = MLPModel()
```

IMO internal dependencies are used so frequently that applying this rule to internal dependencies (e.g. `MLPModel`) would just make the code *harder* to read and for little benefit. A variation on this style is practiced in some places, though, e.g. Google Python style practices a variation where you can import modules but not objects, e.g.

```python
import collections

from garage.tf import models

q = collections.deque(10)
m = models.MLPModel()  # not bad
```

I don't know anyone who practices full paths for both internal and external dependencies, but I'm sure someone does it for some reason:
```python
import collections

import garage.tf.models

q = collections.deque(10)
m = garage.tf.models.MLPModel(m)  # ick
```

The assumption is that internal code will be used more frequently than external code,

with a few notable exceptions, which can be abbreviated with standard notations (e.g. tf, np, torch) to shorten them.

Reply:
on reason behind the google variation is that their codebase is so big that, past a couple parent packages, everything might as well be considered "external" code, even if it's owned by Google (because they all live in a single package namespace).

We don't have this problem, so I'm not sure forcing people to include the module as a stem (e.g. `models.MLPModel()`) really achieves much.

Reply:
one argument in favor of the google style, though, if that you're virtually guaranteed that `models` is defined either in `models.py` or a directory `models/`, whereas `MLPModel` could be anywhere.

Reply:
Yeah, this all sounds about right to me. One note is that I've sometimes seen styles where specific external modules do allow importing classes / functions individually (ironically, collections itself being the most common example). I could see this either way, especially because garage has been "exploded" into other libraries controlled by the rlworkgroup. For example, no one seems keen to write `dowel.logger.log(...)` everywhere. We could probably keep this list fairly short, and it would still be helpful.

On a related note, by convention PyTorch actually has three imports: `torch`, `torch.nn as nn`, and `torch.nn.functional as F`.

Reply:
I will add a list of known exceptions